### ISSUE TITLE
CRD to Source Code Generator

### ISSUE THREAD

**Issue by csviri** [java-operator-sdk/java-operator-sdk#28]
Generate Custom Resource POJO from CRD.

**Comment by csviri** [comment]
see open api generator
https://github.com/OpenAPITools/openapi-generator

**Comment by aegershman** [comment]
apologies if I'm missing something, but is the reverse currently possible? e.g., is there a way to generate the corresponding crd `yml` purely from annotated source code?

What I'm thinking is something similar to using a `//+comment-marker` in golang using `controller-gen`, [as described here in the kubebuilder tooling](https://book.kubebuilder.io/reference/generating-crd.html)

my apologies if I'm looking in the wrong spot for doc / issues, or if I have a misconception of how the crd definition can even be generated

**Comment by csviri** [comment]
Hi @aegershman , np, thx for the question, but unfortunately this is not yet possible. 

We were discussing this also, but implementing other way around, so CRD to Pojo generation should be much easier to implement (thus this is how I see it now, since there are libraries which already do that).

But will create an issue, and will discuss it further probably in close future.

**Comment by adam-sandor** [comment]
Yeah this is a big question whether we should go "contract first" and generate classes from the CRD yaml. Or be "code first" and generate yamls from annotated Java POJOs. For now we are thinking "contract first" to start with, but might support both in the future.

**Comment by csviri** [comment]
yep, I'm fan of contract first.
(If we generate it from POJO is more of a risk, since for example a version change of a underlying library can affect the contract we generate. )

**Comment by psycho-ir** [comment]
I can think of 2 ways to implement this feature. 
1. A CLI application
2. A plugin for the popular build tools like Maven & Gradle.

What I like about Option 1 is that it's build tools agnostic and we still have this possibility to build the plugins for build tools on top of it.

Would be great if you could share your thoughts about this feature anytime you have time.
@adam-sandor @csviri @kirek007  @metacosm

**Comment by psycho-ir** [comment]
I would start with a CLI tool as a separate maven module for the first version.
We can create build tools plugins for the ones have demand later.

**Comment by metacosm** [comment]
> apologies if I'm missing something, but is the reverse currently possible? e.g., is there a way to generate the corresponding crd `yml` purely from annotated source code?

@aegershman Yes, it's currently possible using dekorate.io where you get your CRD yaml from annotated java source.

**Comment by metacosm** [comment]
> yep, I'm fan of contract first.
> (If we generate it from POJO is more of a risk, since for example a version change of a underlying library can affect the contract we generate. )

The yaml-to-pojo approach doesn't fit how Java developers would go about developing operators, in my opinion. As a Java developer, my contract is my POJO and I'd like to avoid dealing with messy yaml files as much as possible and in particular, I really don't want to create a crd.yml by hand (which might take a substantial amount of time) to generate POJO files that I can write in 2 minutes.

**Comment by metacosm** [comment]
To be clear, I'm not saying that such a tool wouldn't be interesting *at some point*. I'm just saying that I don't see it fitting the target audience's workflow all that well. People I've talked to who already have go operators and are interested to migrating to a java-based operator would much rather use the go struct to java classes approach used by Fabric8 client than generate Java classes from YAML.

**Comment by csviri** [comment]
@metacosm @psycho-ir I think both appraoches - the contract first and generating CRD definitions - has its  benefits. Maybe it's just me, thinking in terms of contract first, since this is usually how it's (how I did) done for  APIs with OpenAPI/Swagger. (Also with SOAP long time ago)

To be honest I as a Java developer would go at this point of time as contract first if there is possibility. But don't think there is consesus in this. Also we could support both.

**Comment by csviri** [comment]
> People I've talked to who already have go operators and are interested to migrating to a java-based operator would much rather use the go struct to java classes approach used by Fabric8 client than generate Java classes from YAML.

This sounds, interesting, I would though if they already have yaml's its easyer to just grap it and generate java files. But if there is already a known demand for this, it makes sense to start with this approach.

**Comment by metacosm** [comment]
> @metacosm @psycho-ir I think both appraoches - the contract first and generating CRD definitions - has its benefits. Maybe it's just me, thinking in terms of contract first, since this is usually how it's done (how I did) for APIs with OpenAPI/Swagger. (Also with SOAP long time ago)

Why wouldn't your POJO be your contract? To me, writing a yaml file is hell, especially when you have to deal with the intricacies of writing the schema, compared to writing a simple POJO and get the yml generated without me having to do anything 😄 

> Also I don't see why we could not support both.

Definitely.

**Comment by csviri** [comment]
> Why wouldn't your POJO be your contract? To me, writing a yaml file is hell, especially when you have to deal with the intricacies of writing the schema, compared to writing a simple POJO and get the yml generated without me having to do anything

I agree that writing open api scheam validation rules for CRD in yaml is like salt in the wound,  but fine tuning pojo-s with detailed validation annotation is not a piece of cake either :)

I think POJOs are the class representation of the contract. But at the end, the contract is what we generate from pojos, and pass to K8S.   
So the classic argument against the the generating from POJO, is that if the underlying implementation of generator changes, the result yaml can change, what in this case is unpleasant side effect. On the other hand if the POJOs are generated differently we see that instantly on compile time.

**Comment by metacosm** [comment]
Presumably, though, you're not re-applying your CRD definition every time you rebuild your operator…

**Comment by psycho-ir** [comment]
> As a Java developer, my contract is my POJO and I'd like to avoid dealing with messy yaml files as much as possible and in particular, I really don't want to create a crd.yml by hand (which might take a substantial amount of time) to generate POJO files that I can write in 2 minutes.

Actually contract is a document between more than a **party**, it's true that one side of the contract is a Java developer in our case but there is no guarantee that other side understands this protocol as well, that's why I believe the contract needs to be language agnostic. 
In addition to that, there are plenty of data that makes sense to have in the definition but not really in the implementation. Even right now in the examples here or many other projects the custom resource struct/class is not fully representative of the entire definition that the CRD has, which makes sense as they don't have an influence on the behaviour and can change later on or in runtime.

**Comment by psycho-ir** [comment]
> To be clear, I'm not saying that such a tool wouldn't be interesting _at some point_. I'm just saying that I don't see it fitting the target audience's workflow all that well. People I've talked to who already have go operators and are interested to migrating to a java-based operator would much rather use the go struct to java classes approach used by Fabric8 client than generate Java classes from YAML.

I believe have the YAML files too, right? why not feeding the javaoperatorsdk with those yaml files and get the custom resource classes? 
IMHO by doing that this way it doesn't matter if their current operator is implemented by Go operator or Kopf or any other framework.

**Comment by metacosm** [comment]
> > As a Java developer, my contract is my POJO and I'd like to avoid dealing with messy yaml files as much as possible and in particular, I really don't want to create a crd.yml by hand (which might take a substantial amount of time) to generate POJO files that I can write in 2 minutes.
> 
> 
> 
> Actually contract is a document between more than a **party**, it's true that one side of the contract is a Java developer in our case but there is no guarantee that other side understands this protocol as well, that's why I believe the contract needs to be language agnostic. 


The "official" contract is still the CRD yaml… it's just that for a Java developer it makes more sense to go from Java to yaml than the reverse.

**Comment by adam-sandor** [comment]
I think what's important here is that implementing class generation from CRD doesn't exclude implementing it the other way around. My guess is there would be demand for both features.

**Comment by aegershman** [comment]
Thanks @metacosm I will try out `dekorate` for crd manifest generation.

This is just from my perspective, but I will share. Before I do, I should say that I'm appreciative of OSS / specifically the java-operator-sdk and the software you all provide / cater to us users, _for free_, so I don't mean for this to sound rude or entitled. Sometimes text translates to being a little more rude than intended, so I apologies in advance if it sounds more confrontational than intended:

>Yeah this is a big question whether we should go "contract first" and generate classes from the CRD yaml. Or be "code first" and generate yamls from annotated Java POJOs. For now we are thinking "contract first" to start with, but might support both in the future.

I feel that going from `Java -> CRD manifest yaml` and `CRD manifest yml -> Java` are separate problems altogether rather than "inversions" of each other. As in, I feel like this isn't a "which to do first" question. Even if you could take a crd yaml manifest and turn it into source code, it doesn't reduce the burdens of having to write a crd yaml manifest by hand in the first place, which is tedious and error-prone; it gets unruly quickly. If we look at other CRDs out there, such as [`jetstack/cert-manager`](https://github.com/jetstack/cert-manager/releases/download/v1.1.0/cert-manager.crds.yaml), those files get into thousands of lines of code quickly. Being able to convert those CRD manifest ymls into source code doesn't change the difficulty of writing the CRD manifest yml in the first place. In fact, it may make certain things worse, because then if I introduce any changes to the generated Java source, it'll be overwritten the next time the CRD manifest yaml generation runs again, and require picking-and-choosing lines of code changes.

This issue might not be the place for it (I feel https://github.com/java-operator-sdk/java-operator-sdk/issues/110 would be a better place, so, apologies for continuing this here), but I will say this because I think it's valuable to differentiate the intended use-case and value of each "pathway"-- I feel it'd better serve most folks to use Java as the "source of truth" and generate things off of it. Generating the crd manifest reduces the burdens of specialized knowledge and hand-crafted toil required to write a crd manifest. Especially if it could do things like have JSR-303 validation constraints turned into `openapiv3validation` constraints, or simultaneously generating a CRD "autocomplete schema" config file for IDEs, e.g. https://github.com/Azure/vscode-kubernetes-tools/pull/583). Or be able to turn `javadoc` on a class into a CRD's `description` field, printer column configuration, etc.

Not to say it doesn't have a place at all. I'd be curious to hear about people/teams who would get enough benefit from going from `crd yaml -> source code` regularly.

**TL;DR, to put this another way; as a user** I would expect the `java-operator-sdk` to be able to take annotated Java source code and turn it into a full CRD yaml manifest ready for distribution, but I would not expect the `java-operator-sdk` to take on transforming a full crd yaml manifest into source code targets of multiple languages (Java being one of them). 

This is just my perspective of course, I doubt I have the same visibility into user's needs that you all do, and it's appreciated what goes into the java-operator-sdk in any case 👍  Anyhow, apologies, this all probably should have gone in this issue over here: https://github.com/java-operator-sdk/java-operator-sdk/issues/110 hopefully this doesn't clutter the issue body too badly

**Comment by metacosm** [comment]
Thanks for the kind word, @aegershman! Let us know how things go with dekorate and what things could further be improved. I will try to add dekorate support here as well so that the only thing that would be needed to generate the CRD would be to add the appropriate dekorate jar on the path.

**Comment by wtrocki** [comment]
Worth mentioning that [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) (defacto standard for golang operators) uses codefirst approach. This will simplify migration from golang operators. 

Many folks already have golang operators and hitting some issues around monitoring and early ecosystem.
Java seems to be the best language that offers mature libraries and implementations.. but migration on it's own should be quite simple.

## Hack for now

I was looking for simplification of migration with some common tools.
The best approach I found so far would be to:

- Migrate all example CR's to JSON
- Use JSON to POJO tool https://json2csharp.com/json-to-pojo
- Add rest manually

**Comment by metacosm** [comment]
We probably will wait for: https://github.com/kubernetes/enhancements/pull/2310

**Comment by andreaTP** [comment]
For some reasons I completely missed this issue, some initial progress in this direction have been done in the `kubernetes-client`:

https://github.com/fabric8io/kubernetes-client/issues/3642
https://github.com/fabric8io/kubernetes-client/pull/3693

(I'm a fan of contract first too @csviri 😛  )

**Comment by csviri** [comment]
Closing this issue, since it is for fabric8 client.

**Comment by andreaTP** [comment]
For reference, a Java generator is now part of the `fabric8` `kubernetes-client` project:
https://github.com/fabric8io/kubernetes-client/tree/master/java-generator
